---
title: "Thaw ponds: mapping reads on Pfam HMMs related to carbon degradation and differential abundance analysis"
author: "Domenico Simone"
date: "August 3rd, 2018"
#output: html_document
output:
  html_document:
    code_folding: hide
    toc: true
    toc_float:
      collapsed: false
      smooth_scroll: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, collapse = TRUE)
library(data.table)
library(readr)
library(stringr)
library(tidyverse)
library(vegan)
library(DESeq2)
library(DT)
```

## Methods in brief

Raw reads were trimmed with **sickle**v1.210 (default parameters) and mapped on Pfam HMMs of interest with **metadomain**. Read datasets were splitted in subsets of 500K reads for this analysis and the results combined in this protocol. Data exploration was performed with the **metaMDS** function available through the R **vegan** package, to perform NMDS to compare the distribution of PFAM entries of interest in the analysed samples. Differential abundance (DA) analysis was performed with R package **DESeq2**. DA analyses were carried out by performing pairwise comparisons of tested multiple conditions, namely *old*, *emerge*, *medium* within epilimnion samples, and *epilimnion*, *metalimnion* and *hypolimnion* within old samples. For each DA analysis, two outputs are shown: 1) MA-plot (described in the next section) and 2) a searchable/filterable table with DA results reported for each PFAM entry. Each table is also exported as csv file.

### Note on MA plots

The MA-plot shows the log2 fold changes over the mean of normalized counts, i.e. the average of counts normalized by size factor. The x axis is the average abundance over all samples, the y axis the log2 fold change between the first condition and the second condition *eg* if an MA-plot shows epilimnion vs metalimnion, PFAM entries with a positive log2 fold change are more abundant in epilimnion samples, when compared to metalimnion samples. For each PFAM entry an adjusted p-value is calculated by the DA pipeline. PFAM entries with an adjusted p-value below a threshold (here 0.1, the suggested default) are shown in red. For each MA-plot, the related data table is reported in a searchable/filterable format, where the aforementioned adjusted p-value is reported in the last column. This table might turn useful for DA analyses returning several differentially abundant PFAM entries.

### Functions used in this report (click "Code" on the right to expand)

```{r}
# code to read all metadomain results
getBigTable <- function(i){
  small.table <- read.table(i, header = FALSE, stringsAsFactors = FALSE)
  colnames(small.table) <- c("PFAM_ID", "PFAM_length", "n_reads", "PFAM_cov", "PFAM_status")
  # i is eg ../metadomain_1M/A-1_S21_L005.0.1.fasta.metadomain
  sample_name <- strsplit(basename(i), "_")[[1]][1]
  p <- data.frame(sample=c(sample_name), stringsAsFactors = FALSE)
  small.table <- cbind(p, small.table)
}

getBigTable.pfam <- function(i){
  small.table <- read.table(i, header = FALSE, stringsAsFactors = FALSE)
  colnames(small.table) <- c("PFAM_ID", "n_reads")
  # i is eg ../metadomain_1M/A-1_S21_L005.0.1.fasta.metadomain
  sample_name <- strsplit(basename(i), "_")[[1]][1]
  p <- data.frame(sample=c(sample_name), stringsAsFactors = FALSE)
  small.table <- cbind(p, small.table)
}

plotMA.significant <- function(df, t=0.1, cex=0.8, main="DE analysis - MA plot"){
  plotMA(df, main=main)
  xx <- df$baseMean
  yy <- df$log2FoldChange
  # Replace NAs in padj with values that you'll never want to consider :-)
  signSubset <- replace(df$padj, is.na(df$padj), 1000) < t
  if(count(signSubset, TRUE) > 0)
    text( xx[signSubset],
          yy[signSubset],
          labels=rownames(df)[signSubset],
          pos=1, cex=cex)
    text( xx[c(TRUE)],
          yy[c(TRUE)],
          labels=c(""))
}
# # pointLabel(x, y, as.character(round(x,5)), offset = 0, cex = .7)
# 
# plotMA.significant.pointlabel <- function(df, t=0.1, cex=0.8, main="DE analysis - MA plot"){
#   plotMA(df, main=main)
#   xx <- df$baseMean
#   yy <- df$log2FoldChange
#   # Replace NAs in padj with values that you'll never want to consider :-)
#   signSubset <- replace(df$padj, is.na(df$padj), 1000) < t
#   if(count(signSubset, TRUE) > 0)
#     pointLabel(xx[signSubset],
#                yy[signSubset],
#                labels=rownames(df)[signSubset],
#                pos=1, cex=cex)
#     # text( xx[signSubset],
#     #       yy[signSubset],
#     #       labels=rownames(df)[signSubset],
#     #       pos=1, cex=cex)
#     text( xx[c(TRUE)],
#           yy[c(TRUE)],
#           labels=c(""))
# }
# 
# plotMA.significant.thigmo <- function(df, t=0.1, cex=0.8, main="DE analysis - MA plot"){
#   plotMA(df, main=main)
#   xx <- df$baseMean
#   yy <- df$log2FoldChange
#   # Replace NAs in padj with values that you'll never want to consider :-)
#   signSubset <- replace(df$padj, is.na(df$padj), 1000) < t
#   if(count(signSubset, TRUE) > 0)
#     thigmophobe.labels(xx[signSubset],
#                yy[signSubset],
#                labels=rownames(df)[signSubset],
#                cex=cex)
#     # pointLabel(xx[signSubset],
#     #            yy[signSubset],
#     #            labels=rownames(df)[signSubset],
#     #            pos=1, cex=cex)
#     # text( xx[signSubset],
#     #       yy[signSubset],
#     #       labels=rownames(df)[signSubset],
#     #       pos=1, cex=cex)
#     text( xx[c(TRUE)],
#           yy[c(TRUE)],
#           labels=c(""))
# }

# since the DESeq2 table has PFAM ids as rownames,
# we need to extract the PFAM ids and add them to the dataframe
# as the first column
DESeq2datatable <- function(DESeq2.data){
  DESeq2.data <- DESeq2.data[which(DESeq2.data$padj < 0.1),]
  datatable(data.frame(cbind(data.frame(PFAM_ID=row.names(DESeq2.data)), DESeq2.data)), rownames = FALSE, filter = "top")
}

# function to write out table DESeq2 table in csv format
DESeq2csv <- function(DESeq2.data, file="DESeq2_data.csv"){
  DESeq2.data <- DESeq2.data[which(DESeq2.data$padj < 0.1),]
  write.table(data.frame(cbind(data.frame(PFAM_ID=row.names(DESeq2.data)), DESeq2.data)), row.names = FALSE, file=file, sep = ",")
}
```

## Mapping of reads on HMM profiles

### Data preparation

```{r}

# Read sample metadata and join it with table with dataset size:

metadata <- read_csv("metadata.csv")
metadata$sample_name <- gsub("_", "-", metadata$sample_code)
colnames(metadata)[(names(metadata) == "size")] <- "age"
rownames(metadata) <- metadata$sample_name

metadata.dataset_size <- read_csv("sample_name_reads")

# add dataset sizes
metadata <- metadata %>%
  left_join(metadata.dataset_size)

# For each sample, read all result files

all_files = list.files("../metadomain_1M", full.names = TRUE)
big.table <- data.frame(rbindlist(lapply(all_files, function(x) getBigTable(paste(x)))))
big.table$n_reads <- as.integer(big.table$n_reads)
big.table$PFAM_length <- as.integer(big.table$PFAM_length)
big.table$PFAM_cov <- as.numeric(big.table$PFAM_cov)

# Group counts by sample and PFAM_ID, **save it**.
all.final.table <- big.table %>% 
  group_by(sample, PFAM_ID, PFAM_length) %>%
  summarise(sum(n_reads)) %>%
  ungroup()

colnames(all.final.table)[4] <- "read_count"

# Convert count table in matrix
all.final.matrix <- all.final.table %>%
  subset(select = -PFAM_length) %>%
  spread(PFAM_ID, read_count)

rownames.all.final.matrix <- all.final.matrix$sample
all.final.matrix <- data.matrix(subset(all.final.matrix, select = -sample)) 
rownames(all.final.matrix) <- rownames.all.final.matrix
```

### Overview of read mappings on PFAM HMMs of interest

```{r, collapse=FALSE, include=TRUE}
data.frame(sample_name=row.names(all.final.matrix), total_mapped_reads=rowSums(all.final.matrix)) %>%
  left_join(metadata.dataset_size) %>%
  mutate("total_mapped_reads (%)"=total_mapped_reads*100/total_reads) %>%
  datatable(rownames = FALSE, filter = "top", options = list(pageLength = 30))
```

### NMDS on normalized data

Perform NMDS on normalized data. Data are normalized by rarefying to the sample with the lowest number of reads mapped on PFAM HMMs of interest and plotted with ggplot.

```{r}
# source: http://geoffreyzahn.com/nmds-example/
min_depth = min(rowSums(all.final.matrix))

# normalize data
all.final.matrix.rarefied <- as.matrix(round(rrarefy(all.final.matrix, min_depth)))

# calculate distance matrix
sample_dist.rarefied <- as.matrix((vegdist(all.final.matrix.rarefied, "bray")))

# perform actual NMDS
NMDS.rarefied <- metaMDS(sample_dist.rarefied)

# build a data frame with NMDS coordinates and metadata
#MDS1 = NMDS$points[,1]
#MDS2 = NMDS$points[,2]
NMDS.rarefied.df = data.frame(MDS1 = NMDS.rarefied$points[,1],
                              MDS2 = NMDS.rarefied$points[,2])
NMDS.rarefied.df$sample_name <- rownames(NMDS.rarefied.df)
NMDS.rarefied.df <- NMDS.rarefied.df %>%
  left_join(metadata)

# Plot with ggplot
ggplot(NMDS.rarefied.df, aes(x=MDS1, y=MDS2, col=age, shape=layer)) +
 geom_point() +
 geom_text(aes(label=sample_name),hjust=-0.15,vjust=0, size=3) +
 #stat_ellipse() +
 theme_bw() +
 labs(title = "NMDS of read mapping on PFAM entries\nrelated to carbon degradation (normalized)")
```

### Differential abundance analysis between ages in epilimnion samples

Generate input data for DESeq2, then create a CountDataSet object.

```{r}
# count table
de.epi.age.matrix <- metadata %>%
  subset(layer == "epilimnion") %>%
  inner_join(all.final.table, by = c("sample_name" = "sample")) %>%
  subset(select = c(sample_name, PFAM_ID, read_count)) %>%
  spread(PFAM_ID, read_count)

colnames.de.epi.age.matrix <- de.epi.age.matrix$sample_name
de.epi.age.matrix <- subset(de.epi.age.matrix, select = -sample_name) %>% t()
colnames(de.epi.age.matrix) <- colnames.de.epi.age.matrix

# metadata
de.epi.age.design <- metadata %>%
  subset(layer == "epilimnion") %>%
  inner_join(all.final.table, by = c("sample_name" = "sample")) %>%
  subset(select = c(sample_name, age)) %>%
  distinct()

de.epi.age.design <- de.epi.age.design[base::order(de.epi.age.design$sample_name),]

rownames.de.epi.age.design <- de.epi.age.design$sample_name
de.epi.age.design <- subset(de.epi.age.design, select = -sample_name)
rownames(de.epi.age.design) <- rownames.de.epi.age.design

# data need to be factor
de.epi.age.design$age <- factor(de.epi.age.design$age)

# Then create a CountDataSet object
ddsFullCountTable <- DESeqDataSetFromMatrix(
countData = de.epi.age.matrix,
colData = de.epi.age.design,
design = ~ age)
```

Run the DESeq2 pipeline and extract results for pairwise comparisons between ages

```{r}
dds <- DESeq(ddsFullCountTable)
res <- results(dds)
res.old_emerge    <- results(dds, contrast = c("age", "old", "emerge"))
res.old_medium    <- results(dds, contrast = c("age", "old", "medium"))
res.emerge_medium <- results(dds, contrast = c("age", "emerge", "medium"))
```

#### Old vs emerging

```{r}
plotMA.significant(res.old_emerge, cex = 0.6, main="DE analysis - MA plot: epilimnion, old vs emerging")
DESeq2datatable(res.old_emerge)
DESeq2csv(res.old_emerge, file="DESeq2.epi.old_vs_emerging.csv")
```

#### Old vs medium

```{r}
plotMA.significant(res.old_medium, cex = 0.6, main="DE analysis - MA plot: epilimnion, old vs medium")
DESeq2datatable(res.old_medium)
DESeq2csv(res.old_medium, file="DESeq2.epi.old_vs_medium.csv")
```

#### Emerging vs medium

```{r}
plotMA.significant(res.emerge_medium, cex = 0.6, main="DE analysis - MA plot: epilimnion, emerging vs medium")
DESeq2datatable(res.emerge_medium)
DESeq2csv(res.emerge_medium, file="DESeq2.epi.emerging_vs_medium.csv")
```

### Differential abundance analysis between layers in old ponds

Generate input data for DESeq2.

```{r}
#### Generate input data for DESeq2:
### count table
de.old.matrix <- metadata %>%
  subset(age == "old") %>%
  inner_join(all.final.table, by = c("sample_name" = "sample")) %>%
  subset(select = c(sample_name, PFAM_ID, read_count)) %>%
  spread(PFAM_ID, read_count)

colnames.de.old.matrix <- de.old.matrix$sample_name
de.old.matrix <- subset(de.old.matrix, select = -sample_name) %>% t()
colnames(de.old.matrix) <- colnames.de.old.matrix

### metadata
de.old.design <- metadata %>%
  subset(age == "old") %>%
  inner_join(all.final.table, by = c("sample_name" = "sample")) %>%
  subset(select = c(sample_name, layer)) %>%
  distinct()

de.old.design <- de.old.design[base::order(de.old.design$sample_name),]

rownames.de.old.design <- de.old.design$sample_name
de.old.design <- subset(de.old.design, select = -sample_name)
rownames(de.old.design) <- rownames.de.old.design

# data need to be factor
de.old.design$layer <- factor(de.old.design$layer)

### Then create a CountDataSet object
de.old.ddsFullCountTable <- DESeqDataSetFromMatrix(
countData = de.old.matrix,
colData = de.old.design,
design = ~ layer)
```

Run the DESeq2 pipeline and extract results for pairwise comparisons between ages.

```{r}
de.old.dds <- DESeq(de.old.ddsFullCountTable)

de.old.res.epi_meta    <- results(de.old.dds, contrast = c("layer", "epilimnion", "meta"))
de.old.res.epi_hypo    <- results(de.old.dds, contrast = c("layer", "epilimnion", "hypo"))
de.old.res.hypo_meta   <- results(de.old.dds, contrast = c("layer", "hypo", "meta"))
```

#### Epilimnion vs metalimnion

```{r}
plotMA.significant(de.old.res.epi_meta, cex = 0.6, main="DA analysis - MA plot: old ponds, epilimnion vs metalimnion")
DESeq2datatable(de.old.res.epi_meta)
DESeq2csv(de.old.res.epi_meta, file="DESeq2.old.epi_vs_meta.csv")
```
catFile=all_samples.counts.all.out
rm -f $catFile
for i in $(ls *.all.out); do
    sampleName=$(echo $i | cut -f1 -d_)
    echo $sampleName
    awk -v sampleName="$sampleName" 'BEGIN{OFS="\t"}{print sampleName, $0}' $i >> $catFile 
done
gzip $catFile

catFile=all_samples.counts.unique.out
rm -f $catFile
for i in $(ls *.unique.out); do
    sampleName=$(echo $i | cut -f1 -d_)
    echo $sampleName
    awk -v sampleName="$sampleName" 'BEGIN{OFS="\t"}{print sampleName, $0}' $i >> $catFile 
done
gzip $catFile
#### Epilimnion vs hypolimnion

```{r, fig.height=8, fig.asp=1.9}
plotMA.significant(de.old.res.epi_hypo, cex = 0.6, main="DA analysis - MA plot: old ponds, epilimnion vs hypolimnion")
DESeq2datatable(de.old.res.epi_hypo)
DESeq2csv(de.old.res.epi_hypo, file="DESeq2.old.epi_vs_hypo.csv")
```

#### Hypolimnion vs metalimnion

```{r}
plotMA.significant(de.old.res.hypo_meta, cex = 0.6, main="DA analysis - MA plot: old ponds, hypolimnion vs metalimnion")
DESeq2datatable(de.old.res.hypo_meta)
DESeq2csv(de.old.res.hypo_meta, file="DESeq2.old.hypo_vs_meta.csv")
```

## Mapping of reads on contigs from coassembly of all samples

Map reads on contigs from coassembly of all samples. Then we check mappings on genes which have at least one of the HMMs of interest. The idea is to consider as match a whole gene even if only a part of it has a match with a HMM profile. If a gene has more than one match in PFAM, counts will be averaged for all HMMs.

### Data preparation

- CDS annotation file from Prodigal
- Read hmmscan result table

```{r}
contigs.cds.annotation <- fread("grep -v '^#' prodigal/thawponds_assembly.cds.out",
           sep = "\t",
           #skip = "#",
           stringsAsFactors = FALSE,
           select = c(1,4,5,7),
           col.names = c("contigID",
                         "start",
                         "end",
                         "strand")) %>% group_by(contigID) %>% 
                                     mutate(count = sequence(n())) %>%
                                     mutate(CDS_ID = paste(contigID, count, sep = "_")) %>%
                                     subset(select = -count) %>% ungroup()

contigs.hmm.pfam <- fread("grep -v '^#' hmmer/split500000/thawponds_assembly.cds.split500000.all.hmmer_pfam.domtblout",
           sep = " ",
           stringsAsFactors = FALSE,
           select = c(1,5,18,19),
           col.names = c("CDS_ID",
                         "PFAM_ID",
                         "hit_start",
                         "hit_end"))
```

Now we want to obtain a gff file with coordinates of hmmscan hits mapped onto the contigs.
Conditionally join to hmmscan hits and extract relevant columns...

```{r}
c <- contigs.hmm.pfam %>%
  left_join(contigs.cds.annotation) %>%
  # start, end : annotations of CDS on the contig
  # hit_start, hit_end : hmmscan result (on the translated CDS)
  mutate(c_start=if_else(strand == "+", start+(hit_start-1)*3, end-hit_end*3+1)) %>%
  mutate(c_end=if_else(strand == "+", start+hit_end*3, end-(hit_start-1)*3)) %>%
  base::cbind(source=c("hmmer_3.2.1"), feature=c("Domain"), score=c("."), frame=c(".")) %>%
  mutate(attribute=paste("PFAM_ID=", PFAM_ID, ";CDS_ID=", CDS_ID, sep="")) %>%
  select(contigID,
         source,
         feature,
         c_start,
         c_end,
         score,
         strand,
         frame,
         attribute)

d <- copy(c)
d$c_start <- format(d$c_start, scientific = FALSE)
d$c_end   <- format(d$c_end,   scientific = FALSE)

write.table(d, 
            sep="\t",
            file="prodigal/thawponds_assembly.cds.hmmer_pfam.gff",
            col.names = FALSE,
            row.names = FALSE,
            quote=FALSE)

remove(d)
```

Export gff to cluster, count reads. 

Count files

```{r}
all_pfam_files = list.files('counts', pattern="*all*", full.names = TRUE)
big.table.pfam <- data.frame(rbindlist(lapply(all_pfam_files, function(x) getBigTable.pfam(paste(x)))))
# big.table$n_reads <- as.integer(big.table$n_reads)
# big.table$PFAM_length <- as.integer(big.table$PFAM_length)
# big.table$PFAM_cov <- as.numeric(big.table$PFAM_cov)

# Group counts by sample and PFAM_ID, **save it**.
all.final.table.pfam <- big.table.pfam %>% 
  group_by(sample, PFAM_ID) %>%
  summarise(sum(n_reads)) %>%
  ungroup()

colnames(all.final.table.pfam)[3] <- "read_count"

# Extract relevant data
all.final.table.pfam.carbon <- all.final.table.pfam %>%
                            separate(PFAM_ID, c("PFAM_ID.noversion", "version"), "\\.") %>%
                            subset(select = -version) %>%
                            inner_join(data.frame(PFAM_ID.noversion=c(colnames(all.final.matrix)))) %>%
                            inner_join()

# Convert count table in matrix
all.final.matrix.pfam.carbon <- all.final.table.pfam.carbon %>%
  spread(PFAM_ID.noversion, read_count)

rownames.all.final.matrix.pfam.carbon <- all.final.matrix.pfam.carbon$sample
all.final.matrix.pfam.carbon <- data.matrix(subset(all.final.matrix.pfam.carbon, select = -sample)) 
rownames(all.final.matrix.pfam.carbon) <- rownames.all.final.matrix.pfam.carbon
all.final.matrix.pfam.carbon <- all.final.matrix.pfam.carbon[-base::nrow(all.final.matrix.pfam.carbon),]
all.final.matrix.pfam.carbon[is.na(all.final.matrix.pfam.carbon)] <- 0

## Obtain relative abundances
# normalize to all reads for each sample
# source: http://geoffreyzahn.com/nmds-example/
min_depth.pfam.carbon = min(rowSums(all.final.matrix.pfam.carbon))

# normalize data
all.final.matrix.pfam.carbon.rarefied <- as.matrix(round(rrarefy(all.final.matrix.pfam.carbon, min_depth.pfam.carbon)))


# Export to file
write.table(t(all.final.matrix.pfam.carbon), file = "thawponds_carbon_degradation_genes_contigs.tsv", sep = "\t", quote = FALSE)
```


```{r}
all_samples.all.counts <- fread(paste0("zcat < ", getwd(), "/all_samples.counts.all.out.gz"),
                                header = FALSE,
                                stringsAsFactors = FALSE)
all_samples.unique.counts <- fread("zcat all_samples.counts.unique.out.gz",
                                header = FALSE,
                                stringsAsFactors = FALSE)
```
